pkg/operator/controller/ingress/deployment_test.go

```diff
@@ -1109,6 +1115,105 @@ func TestDesiredRouterDeploymentClientTLS(t *testing.T) {
 	}
 }
 
+// TestDesiredRouterDeploymentDynamicConfigManager verifies that desiredRouterDeployment
+// returns the expected deployment when DCM featuregate is involved.
+func TestDesiredRouterDeploymentDynamicConfigManager(t *testing.T) {
+	testCases := []struct {
+		name                       string
+		unsupportedConfigOverrides string
+		dcmEnabled                 bool
+		expectedEnv                []envData
+	}{
+		{
+			name:       "featuregate-enabled-default-values",
+			dcmEnabled: true,
+			expectedEnv: []envData{
+				{"ROUTER_HAPROXY_CONFIG_MANAGER", true, "true"},
+				{"ROUTER_MAX_DYNAMIC_SERVERS", true, "1"},
+				{"ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", true, "0"},
+			},
+		},
+		{
+			name:                       "featuregate-enabled-max-servers",
+			unsupportedConfigOverrides: `{"maxDynamicServers":"7"}`,
+			dcmEnabled:                 true,
+			expectedEnv: []envData{
+				{"ROUTER_HAPROXY_CONFIG_MANAGER", true, "true"},
+				{"ROUTER_MAX_DYNAMIC_SERVERS", true, "7"},
+				{"ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", true, "0"},
+			},
+		},
+		{
+			name:                       "featuregate-enabled-dcm-override",
+			unsupportedConfigOverrides: `{"dynamicConfigManager":"false","maxDynamicServers":"7"}`,
+			dcmEnabled:                 true,
+			expectedEnv: []envData{
+				{"ROUTER_HAPROXY_CONFIG_MANAGER", false, ""},
+				{"ROUTER_MAX_DYNAMIC_SERVERS", false, ""},
+				{"ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", false, ""},
+			},
+		},
+		{
+			name:       "featuregate-disabled-default-values",
+			dcmEnabled: false,
+			expectedEnv: []envData{
+				{"ROUTER_HAPROXY_CONFIG_MANAGER", false, ""},
+				{"ROUTER_MAX_DYNAMIC_SERVERS", false, ""},
+				{"ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", false, ""},
+			},
+		},
+		{
+			name:                       "featuregate-disabled-max-servers",
+			unsupportedConfigOverrides: `{"maxDynamicServers":"7"}`,
+			dcmEnabled:                 false,
+			expectedEnv: []envData{
+				{"ROUTER_HAPROXY_CONFIG_MANAGER", false, ""},
+				{"ROUTER_MAX_DYNAMIC_SERVERS", false, ""},
+				{"ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", false, ""},
+			},
+		},
+		{
+			name:                       "featuregate-disabled-dcm-override",
+			unsupportedConfigOverrides: `{"dynamicConfigManager":"true","maxDynamicServers":"7"}`,
+			dcmEnabled:                 false,
+			expectedEnv: []envData{
+				{"ROUTER_HAPROXY_CONFIG_MANAGER", true, "true"},
+				{"ROUTER_MAX_DYNAMIC_SERVERS", true, "7"},
+				{"ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", true, "0"},
```

Comment:
Are we going to allow users the over ride turn this **on** without the feature gate enabled?  I thought we were just going to allow them an over ride to turn it **off** after the feature gate was enabled.

Reply:
I just kept it compatible with the current behavior. Currently (4.17 and before), we can enable DCM using `upsupportedConfigOverride` even though it would be not a wise decision as it has some gaps (which will be fixed in 4.18).

Reply:
Unsupported config override is exactly what the name says.  If you use this, you're unsupported.  It's right there in the name, as explicit as we can make it without descending into absurdity (à la the `unsupported.do-not-use.openshift.io/override-liveness-grace-period-seconds` annotation).  If you use this option, you're overriding the operator and by extension the best judgment of the engineers who wrote the operator.  We're not going to second-guess you or encumber your (or ourselves!) with a bunch of validation or restrictions.  You're on your own, and you better know what you're doing if you use an unsupported config override.